File: delete-and-earn.cs
public class Solution {
    public int DeleteAndEarn(int[] nums) {
         if(nums == null || nums.Length == 0) 
        {
            return 0;
        }
        int max = Int32.MinValue;
        foreach(int num in nums)
        {
            max = Math.Max(max,num);
        }
        int[] dp = new int[max+1];
        foreach(int i in nums)
        {
            dp[i]+=i;
        }
        int skip=0;int take=0;
        for(int j=0;j<dp.Length;j++)
        {
            int temp = Math.Max(skip,take);
            take = skip +dp[j];
            skip = temp;
        }
        return Math.Max(skip,take);
        
    
        
    }
}

File: Delete and Earn.cpp
//740. Delete and Earn

/*
First get maximum value in the array, and get sum of all frequencies of an element

Time Complexity: O(N)
Space Complexity: O(N)

*/

class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {

        
        //using arrays
        int maxE = INT_MIN;
        //get max valued element
        for(int n:nums) maxE = max(maxE, n);
        vector<int> summ(maxE+1); //+1 cause index starts from 0

        //get sum of all freqs
        for(int n:nums) summ[n]+=n;

        int prev = summ[0];
        int curr = summ[1];

        for(int i=2; i<summ.size();i++){
            int temp = curr;
            curr = max(curr, summ[i]+prev);
            prev = temp;
        }

        return curr;
        

        //using hashed maps
        
    }
};